README.md
# ntriples-collection

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Coveralls Status][coveralls-image]][coveralls-url]
[![Dependency Status][depstat-image]][depstat-url]
[![Downloads][download-badge]][npm-url]

> Utility methods for filtering n-triples

Assumptions:
* You are using [node.js n3](https://github.com/RubenVerborgh/N3.js)
* You are filtering small arrays of triples.
* You are mostly interested in the object value.

## Install

```sh
npm i -D ntriples-collection
```

## Usage

```js
import { readNTriplesFile,
   writeNTriplesFile,
   findObjectByPredicate,
   findLocalizedObjectByPredicate,
   findObjectsByPredicate,
  } from "ntriples-collection"
```
### Functions

<dl>
<dt><a href="#readNTriplesFile">readNTriplesFile(filename, function)</a></dt>
<dd><p>Reads a n-triples file and converts it to an array of triples</p>
</dd>
<dt><a href="#writeNTriplesFile">writeNTriplesFile(filename, triples, function)</a></dt>
<dd><p>Saves an array of triples in a n-triples file</p>
</dd>
<dt><a href="#findObjectByPredicate">findObjectByPredicate(triples, predicate, defaultValue)</a> ⇒ <code>string</code></dt>
<dd><p>Finds the first object value based on the predicate</p>
</dd>
<dt><a href="#findLocalizedObjectByPredicate">findLocalizedObjectByPredicate(triples, predicate, language, altLangs, defaultValue)</a> ⇒ <code>string</code></dt>
<dd><p>Finds the first object value based on the predicate and the language</p>
</dd>
<dt><a href="#findObjectsByPredicate">findObjectsByPredicate(triples, predicate)</a> ⇒ <code>array</code></dt>
<dd><p>Finds all object values based on the predicate</p>
</dd>
</dl>

<a name="readNTriplesFile"></a>

### readNTriplesFile(filename, function)
Reads a n-triples file and converts it to an array of triples

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| filename | <code>string</code> | the n-triples filename |
| function | <code>callback</code> | callback |

**Example**  
```js
// [
{ graph: "",
object: '"Flower corp"@en',
predicate: 'http://purl.org/dc/elements/1.1/publisher',
subject: 'http://www.site.org/version/123/'
},
]
readNTriplesFile('flowers.nt', (err, triples) => {
 console.log(triples);
});
```
<a name="writeNTriplesFile"></a>

### writeNTriplesFile(filename, triples, function)
Saves an array of triples in a n-triples file

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| filename | <code>string</code> | the n-triples filename |
| triples | <code>array</code> | an array of triples objects |
| function | <code>callback</code> | callback |

**Example**  
```js
// {count: 1}
 const triples = [
{ graph: "",
object: '"Flower corp"@en',
predicate: 'http://purl.org/dc/elements/1.1/publisher',
subject: 'http://www.site.org/version/123/'
},
]
writeNTriplesFile('flowers.nt', triples, (err, triples) => {
 console.log(triples);
});
```
<a name="findObjectByPredicate"></a>

### findObjectByPredicate(triples, predicate, defaultValue) ⇒ <code>string</code>
Finds the first object value based on the predicate

**Kind**: global function  
**Returns**: <code>string</code> - the string, integer, float, boolean, moment representing the literal value  

| Param | Type | Description |
| --- | --- | --- |
| triples | <code>array</code> | an array of triples objects (subject is ignored) |
| predicate | <code>string</code> | the uri representing the predicate |
| defaultValue | <code>object</code> | the object/string to return if null |

**Example**  
```js
// returns Amadeus
findObjectByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator')
```
<a name="findLocalizedObjectByPredicate"></a>

### findLocalizedObjectByPredicate(triples, predicate, language, altLangs, defaultValue) ⇒ <code>string</code>
Finds the first object value based on the predicate and the language

**Kind**: global function  
**Returns**: <code>string</code> - the string, integer, float, boolean, moment representing the literal value  

| Param | Type | Description |
| --- | --- | --- |
| triples | <code>array</code> | an array of triples objects (subject is ignored) |
| predicate | <code>string</code> | the uri representing the predicate |
| language | <code>string</code> | the requested language |
| altLangs | <code>array</code> | an array of alternative languages (max 2) |
| defaultValue | <code>object</code> | the object/string to return if null |

**Example**  
```js
// returns Amadeus
findLocalizedObjectByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator', 'fr', ['en'])
```
<a name="findObjectsByPredicate"></a>

### findObjectsByPredicate(triples, predicate) ⇒ <code>array</code>
Finds all object values based on the predicate

**Kind**: global function  
**Returns**: <code>array</code> - of string, integer, float, boolean, moment representing the literal values  

| Param | Type | Description |
| --- | --- | --- |
| triples | <code>array</code> | an array of triples objects (subject is ignored) |
| predicate | <code>string</code> | the uri representing the predicate |

**Example**  
```js
// returns [Amadeus, Bach]
findObjectsByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator')
```


## License

MIT © [Olivier Huin](http://github.com/flarebyte)

[npm-url]: https://npmjs.org/package/ntriples-collection
[npm-image]: https://img.shields.io/npm/v/ntriples-collection.svg?style=flat-square

[travis-url]: https://travis-ci.org/flarebyte/ntriples-collection
[travis-image]: https://img.shields.io/travis/flarebyte/ntriples-collection.svg?style=flat-square

[coveralls-url]: https://coveralls.io/r/flarebyte/ntriples-collection
[coveralls-image]: https://img.shields.io/coveralls/flarebyte/ntriples-collection.svg?style=flat-square

[depstat-url]: https://david-dm.org/flarebyte/ntriples-collection
[depstat-image]: https://david-dm.org/flarebyte/ntriples-collection.svg?style=flat-square

[download-badge]: http://img.shields.io/npm/dm/ntriples-collection.svg?style=flat-square


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Changelog

- [v1.0.0](#v100)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### v1.0.0

 * ntriples-collection initial commit.


test/index.js
import test from 'tape';
import path from 'path';
import fs from 'fs';
import _ from 'lodash';
import { readNTriplesFile,
   writeNTriplesFile,
   findObjectByPredicate,
   findLocalizedObjectByPredicate,
   findObjectsByPredicate,
  } from '../src';

const fixturesDir = path.resolve(__dirname, 'fixtures');
const tmpDir = path.resolve(__dirname, 'tmp');

if (!fs.existsSync(tmpDir)) {
  fs.mkdirSync(tmpDir);
}

test('ntriplesCollection should read n-triples file', (t) => {
  t.plan(9);
  const expected = [
    { object: '"Publisher Alpha"', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/123/' },
    { object: '"Creator Alpha"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/123/' },
    { object: '"Publisher Beta"', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/124/' },
    { object: '"Creator Beta"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/124/' },
    { object: '"Dave Beckett"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/125/' },
    { object: '"Art Barstow"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/125/' },
    { object: 'http://www.w3.org/', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/125/' },
  ];
  readNTriplesFile(`${fixturesDir}/reference.nt`, (err, data) => {
    t.equal(err, null, 'no error');
    t.equal(_.size(data), 7, 'size of array');
    t.deepEqual(data[0], expected[0], 'Publisher Alpha');
    t.deepEqual(data[1], expected[1], 'Creator Alpha');
    t.deepEqual(data[2], expected[2], 'Publisher Beta');
    t.deepEqual(data[3], expected[3], 'Creator Beta');
    t.deepEqual(data[4], expected[4], 'Dave Beckett');
    t.deepEqual(data[5], expected[5], 'Art Barstow');
    t.deepEqual(data[6], expected[6], 'http://www.w3.org/');
  });
});

test('ntriplesCollection should write n-triples file', (t) => {
  t.plan(3);
  const expected = fs.readFileSync(`${fixturesDir}/write-reference.nt`, 'utf8');
  const graph = '/only-for-quad';
  const triples = [
    { graph, object: '"Publisher Alpha"', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/123/' },
    { graph, object: '"Creator Alpha"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/123/' },
    { graph, object: '"Publisher Beta"', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/124/' },
    { graph, object: '"Creator Beta"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/124/' },
    { graph, object: '"Dave Beckett"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/125/' },
    { graph, object: '"Art Barstow"', predicate: 'http://purl.org/dc/elements/1.1/creator', subject: 'http://www.site.org/version/125/' },
    { graph, object: 'http://www.w3.org/', predicate: 'http://purl.org/dc/elements/1.1/publisher', subject: 'http://www.site.org/version/125/' },
  ];
  writeNTriplesFile(`${tmpDir}/write-example.nt`, triples, (err, results) => {
    t.equal(err, null, 'no error');
    t.equal(results.count, 7, 'count of triples');
    const actual = fs.readFileSync(`${tmpDir}/write-example.nt`, 'utf8');
    t.deepEqual(actual, expected, 'same n-triples');
  });
});

test('ntriplesCollection should find object by predicate', (t) => {
  t.plan(11);
  const graph = '/only-for-quad';
  const subject = 'http://www.site.org/subject/123/';
  const triples = [
    { object: '"A"', predicate: 'http://w.org/a', subject, graph },
    { object: '"B"', predicate: 'http://w.org/b', subject, graph },
    { object: '"C"', predicate: 'http://w.org/c', subject },
    { object: '"D"@en', predicate: 'http://w.org/d', subject, graph },
    { object: '"D2"', predicate: 'http://w.org/d', subject, graph },
    { object: '"E"', predicate: 'http://w.org/e', subject },
    { object: '"D3"', predicate: 'http://w.org/d', subject, graph },
    { object: '"3"^^http://www.w3.org/2001/XMLSchema#integer', predicate: 'http://w.org/g', subject },
    { object: '"3.2"^^http://www.w3.org/2001/XMLSchema#float', predicate: 'http://w.org/h', subject, graph },
    { object: '"true"^^http://www.w3.org/2001/XMLSchema#boolean', predicate: 'http://w.org/i', subject },
    { object: '"2017-04-01T10:45:54Z"^^http://www.w3.org/2001/XMLSchema#dateTime', predicate: 'http://w.org/j', subject, graph },
    { object: '"2017-03-01"^^http://www.w3.org/2001/XMLSchema#date', predicate: 'http://w.org/k', subject },
  ];
  const actualB = findObjectByPredicate(triples, 'http://w.org/b');
  const actualC = findObjectByPredicate(triples, 'http://w.org/c');
  const actualD = findObjectByPredicate(triples, 'http://w.org/d');
  const actualE = findObjectByPredicate(triples, 'http://w.org/e');
  const actualG = findObjectByPredicate(triples, 'http://w.org/g');
  const actualH = findObjectByPredicate(triples, 'http://w.org/h');
  const actualI = findObjectByPredicate(triples, 'http://w.org/i');
  const actualJ = findObjectByPredicate(triples, 'http://w.org/j');
  const actualK = findObjectByPredicate(triples, 'http://w.org/k');
  const actualUnknown = findObjectByPredicate(triples, 'http://w.org/none');
  const actualUnknownDef = findObjectByPredicate(triples, 'http://w.org/none', 'NULL');

  t.equal(actualUnknown, null, 'Unknown');
  t.equal(actualUnknownDef, 'NULL', 'Unknown with default');
  t.equal(actualB, 'B', 'B');
  t.equal(actualC, 'C', 'C');
  t.equal(actualD, 'D', 'D');
  t.equal(actualE, 'E', 'E');
  t.equal(actualG, 3, 'Integer');
  t.equal(actualH, 3.2, 'Float');
  t.ok(actualI, 'boolean');
  t.equal(actualJ.format('YYYY/mm'), '2017/45', 'DateTime');
  t.equal(actualK.format('YYYY/MM'), '2017/03', 'Date');
});

test('ntriplesCollection should find localized object by predicate', (t) => {
  t.plan(10);
  const graph = '/only-for-quad';
  const subject = 'http://www.site.org/subject/123/';
  const triples = [
    { object: '"A"', predicate: 'http://w.org/a', subject },
    { object: '"A-en"@en', predicate: 'http://w.org/a', subject, graph },
    { object: '"A-fr"@fr', predicate: 'http://w.org/a', subject },
    { object: '"A-es"@es', predicate: 'http://w.org/a', subject, graph },
    { object: '"B"', predicate: 'http://w.org/b', subject },
  ];

  const find = (lang, langs, defval) => findLocalizedObjectByPredicate(triples, 'http://w.org/a', lang, langs, defval);

  t.equal(find('fr'), 'A-fr', 'fr');
  t.equal(find(''), 'A', 'no lang');
  t.equal(find('de'), null, 'de');
  t.equal(find('de', [], 'NULL'), 'NULL', 'de NULL');
  t.equal(find('de', ['en']), 'A-en', 'de then en');
  t.equal(find('de', ['jp', '']), 'A', 'de then jp then no localised');
  t.equal(find('de', ['jp']), null, 'de then jp');
  t.equal(find('de', ['jp'], 'NULL'), 'NULL', 'de then jp NULL');
  t.equal(find('de', ['jp', 'it']), null, 'de then jp then it');
  t.equal(find('de', ['jp', 'it'], 'NULL'), 'NULL', 'de then jp then it NULL');
});

test('ntriplesCollection should find localized object by predicate', (t) => {
  t.plan(2);
  const graph = '/only-for-quad';
  const subject = 'http://www.site.org/subject/123/';
  const triples = [
    { object: '"A"', predicate: 'http://w.org/a', subject, graph },
    { object: '"A-en"@en', predicate: 'http://w.org/a', subject },
    { object: '"A-fr"@fr', predicate: 'http://w.org/a', subject, graph },
    { object: '"A-es"@es', predicate: 'http://w.org/a', subject },
    { object: '"B"', predicate: 'http://w.org/b', subject, graph },
  ];

  const actual = findObjectsByPredicate(triples, 'http://w.org/a');
  const expected = ['A', 'A-en', 'A-fr', 'A-es'];
  t.deepEqual(actual, expected, 'a list of A');
  t.deepEqual(findObjectsByPredicate(triples, 'http://w.org/z'), [], 'unknown');
});


src/index.js
import _ from 'lodash';
import _S from 'string';
import moment from 'moment';
import fs from 'fs';
import { Util, Parser, Writer } from 'n3';

const isTriple = str => _S(str).contains('<');

const omitGraph = t => _.pick(t, ['subject', 'predicate', 'object']);
const withoutGraph = triples => _.map(triples, omitGraph);

const normalizeValue = (value, defaultValue) => {
  if (_.isNil(value)) {
    return defaultValue;
  }
  if (Util.isLiteral(value)) {
    const litType = Util.getLiteralType(value);
    const litValue = Util.getLiteralValue(value);
    switch (litType) {
      case 'http://www.w3.org/2001/XMLSchema#string': return litValue;
      case 'http://www.w3.org/2001/XMLSchema#anyURI': return litValue;
      case 'http://www.w3.org/2001/XMLSchema#integer': return _S(litValue).toInt();
      case 'http://www.w3.org/2001/XMLSchema#nonPositiveInteger': return _S(litValue).toInt();
      case 'http://www.w3.org/2001/XMLSchema#negativeInteger': return _S(litValue).toInt();
      case 'http://www.w3.org/2001/XMLSchema#nonNegativeInteger': return _S(litValue).toInt();
      case 'http://www.w3.org/2001/XMLSchema#positiveInteger': return _S(litValue).toInt();
      case 'http://www.w3.org/2001/XMLSchema#float': return _S(litValue).toFloat();
      case 'http://www.w3.org/2001/XMLSchema#boolean': return _S(litValue).toBool();
      case 'http://www.w3.org/2001/XMLSchema#dateTime': return moment(litValue, moment.ISO_8601);
      case 'http://www.w3.org/2001/XMLSchema#date': return moment(litValue, moment.ISO_8601);
      case 'http://www.w3.org/2001/XMLSchema#time': return litValue;
      default: return litValue;
    }
  }
  if (Util.isIRI(value)) {
    return value;
  }
  return value;
};

const normalizeValues = (values, defaultValue) =>
  _.map(values, v => normalizeValue(v, defaultValue));

const asSemanticValue = (value) => {
  const literalValue = normalizeValue(value, null);
  const literalType = Util.getLiteralType(value);
  const literalLanguage = Util.getLiteralLanguage(value);
  return { value, literalValue, literalType, literalLanguage };
};

const asSemanticValues = values => _.map(values, asSemanticValue);

/**
 * Reads a n-triples file and converts it to an array of triples
 * @param {string} filename - the n-triples filename
 * @param {callback} function - callback
 * @example
 * // [
 * { graph: "",
 * object: '"Flower corp"@en',
 * predicate: 'http://purl.org/dc/elements/1.1/publisher',
 * subject: 'http://www.site.org/version/123/'
 * },
 * ]
 * readNTriplesFile('flowers.nt', (err, triples) => {
 *  console.log(triples);
 * });
 */
export function readNTriplesFile(filename, callback) {
  const n3parser = new Parser();
  const n3parse = str => _.head(n3parser.parse(str));

  fs.readFile(filename, 'utf8', (err, data) => {
    if (err) {
      callback(err);
    } else {
      const lines = _.filter(_S(data).lines(), isTriple);
      const triples = _.map(lines, n3parse);
      callback(null, withoutGraph(triples));
    }
  });
}

/**
 * Saves an array of triples in a n-triples file
 * @param {string} filename - the n-triples filename
 * @param {array} triples - an array of triples objects
 * @param {callback} function - callback
 * @example
 * // {count: 1}
 *  const triples = [
 * { graph: "",
 * object: '"Flower corp"@en',
 * predicate: 'http://purl.org/dc/elements/1.1/publisher',
 * subject: 'http://www.site.org/version/123/'
 * },
 * ]
 * writeNTriplesFile('flowers.nt', triples, (err, triples) => {
 *  console.log(triples);
 * });
 */
export function writeNTriplesFile(filename, triples, callback) {
  const fstream = fs.createWriteStream(filename);
  const n3writer = new Writer(fstream, { format: 'N-Triples' });
  n3writer.addTriples(withoutGraph(triples));
  n3writer.end((error) => {
    const count = _.size(triples);
    const e = _.isNil(error) ? null : { error, filename };
    callback(e, { count });
  });
}

/**
 * Finds the first object value based on the predicate
 * @param {array} triples - an array of triples objects (subject is ignored)
 * @param {string} predicate - the uri representing the predicate
 * @param {object} defaultValue - the object/string to return if null
 * @example
 * // returns Amadeus
 * findObjectByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator')
 * @return {string} the string, integer, float, boolean, moment representing the literal value
 */
export function findObjectByPredicate(triples, predicate, defaultValue = null) {
  const value = _.get(_.find(triples, { predicate }), 'object');
  return normalizeValue(value, defaultValue);
}

/**
 * Finds the first object value based on the predicate and the language
 * @param {array} triples - an array of triples objects (subject is ignored)
 * @param {string} predicate - the uri representing the predicate
 * @param {string} language - the requested language
 * @param {array} altLangs - an array of alternative languages (max 2)
 * @param {object} defaultValue - the object/string to return if null
 * @example
 * // returns Amadeus
 * findLocalizedObjectByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator', 'fr', ['en'])
 * @return {string} the string, integer, float, boolean, moment representing the literal value
 */
export function findLocalizedObjectByPredicate(triples, predicate, language, altLangs,
  defaultValue = null) {
  const values = _.map(_.filter(triples, { predicate }), 'object');
  if (_.isEmpty(values)) {
    return defaultValue;
  }

  const semanticValues = asSemanticValues(values);
  const altLang = _.nth(altLangs);
  const altLang2 = _.nth(altLangs, 1);
  const value = _.find(semanticValues, { literalLanguage: language });
  if (_.isNil(altLang) || !_.isNil(value)) {
    return _.get(value, 'literalValue', defaultValue);
  }
  const altValue = _.find(semanticValues, { literalLanguage: altLang });
  if (_.isNil(altLang2) || !_.isNil(altValue)) {
    return _.get(altValue, 'literalValue', defaultValue);
  }
  const altValue2 = _.find(semanticValues, { literalLanguage: altLang2 });
  return _.get(altValue2, 'literalValue', defaultValue);
}

/**
 * Finds all object values based on the predicate
 * @param {array} triples - an array of triples objects (subject is ignored)
 * @param {string} predicate - the uri representing the predicate
 * @example
 * // returns [Amadeus, Bach]
 * findObjectsByPredicate(triples, 'http://purl.org/dc/elements/1.1/creator')
 * @return {array} of string, integer, float, boolean, moment representing the literal values
 */
export function findObjectsByPredicate(triples, predicate) {
  const values = _.map(_.filter(triples, { predicate }), 'object');
  return normalizeValues(values);
}
